app/wasm/message_plugin.go

```diff
@@ -0,0 +1,201 @@
+package wasm
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+
+	wasmbindings "github.com/osmosis-labs/osmosis/v7/app/wasm/bindings"
+	gammkeeper "github.com/osmosis-labs/osmosis/v7/x/gamm/keeper"
+	gammtypes "github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
+func CustomMessageDecorator(gammKeeper *gammkeeper.Keeper, bank *bankkeeper.BaseKeeper) func(wasmkeeper.Messenger) wasmkeeper.Messenger {
+	return func(old wasmkeeper.Messenger) wasmkeeper.Messenger {
+		return &CustomMessenger{
+			wrapped:    old,
+			bank:       bank,
+			gammKeeper: gammKeeper,
+		}
+	}
+}
+
+type CustomMessenger struct {
+	wrapped    wasmkeeper.Messenger
+	bank       *bankkeeper.BaseKeeper
+	gammKeeper *gammkeeper.Keeper
+}
+
+var _ wasmkeeper.Messenger = (*CustomMessenger)(nil)
+
+func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
```

Comment:
This is the entry point to handle messages. 
If you add another variant to `OsmosisMsg`, you can add the handler here.